thrift.js

```diff
@@ -58,6 +59,46 @@ var messageExceptionTypesDef = require('./message').exceptionTypesDef;
 var validThriftIdentifierRE = /^[a-zA-Z_][a-zA-Z0-9_\.]+$/;
 
 function Thrift(options) {
+    this._init(options);
+
+    if (this.asyncLoad) {
+        return;
+    }
+
+    this._parse(this.filename, this.allowIncludeAlias);
+    this._compileAndLink();
+}
+
+// Alternative constructor allowing for asynchronous source loading.
+Thrift.load = function load(options, load, cb) {
+    if (!options) {
+        return cb(Error('options required'));
+    } else if (typeof options !== 'object') {
+        return cb(Error('options must be object'));
+    } else if (!load) {
+        return cb(Error("Thrift.load must be passed a 'load' function as second argument"));
+    }
+    options.fs = {load: load};
+    var thrift;
+    try {
+        thrift = new Thrift(options);
+    } catch (err) {
+        return cb(err, undefined);
+    }
+    thrift._asyncParse(thrift.filename, thrift.allowIncludeAlias, function (err) {
+        if (err) {
+            return cb(err, undefined);
+        }
+        try {
+            thrift._compileAndLink();
+        } catch (err) {
+            return cb(err, undefined);
+        }
```

Comment:
Out of scope for this change, but we generally prefer not to use try/throw since try/catch deoptimizes the surrounding function in v8 and muddies stack traces and heap dumps for legitimately unrecoverable errors like TypeError, ReferenceError, and RangeError.

Perhaps there’s a future where we can convert all of our user errors like the one for bad include directives into errors that do not throw.

I’m okay with keeping this as is because you will clearly want to recover from errors due to invalid Thrift IDL. The PEG generated code clearly throws exceptions, but could be isolated in the parse phase. I believe the link errors are the only cases that throw for invalid Thrift IDL outside of parsing.

And, clearly, changing an exception throwing to error annotating API would be a breaking change, so improving this would need to be orthogonal to making browser loading just work.